src/main/flight/alt_hold.c

```diff
@@ -90,65 +89,66 @@ void altHoldProcessTransitions(void) {
-    } else if (rcThrottle > highThreshold) {
-        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    // User can adjust the target altitude with throttle, but only when
+    // - throttle is outside deadband, and 
+    // - throttle is not low (zero), and 
+    // - deadband is not configured to zero
+
+    float stickFactor = 0.0f;
+
+    if (altHold.allowStickAdjustment && calculateThrottleStatus() != THROTTLE_LOW) {
+        const float rcThrottle = rcCommand[THROTTLE];
+        const float lowThreshold = autopilotConfig()->hover_throttle - altHold.deadband * (autopilotConfig()->hover_throttle - PWM_RANGE_MIN);
+        const float highThreshold = autopilotConfig()->hover_throttle + altHold.deadband * (PWM_RANGE_MAX - autopilotConfig()->hover_throttle);
+
+        if (rcThrottle < lowThreshold) {
+            stickFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+        } else if (rcThrottle > highThreshold) {
+            stickFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+        }
     }
-    // if failsafe is active, and we get here, we are in failsafe landing mode
+
+    // if failsafe is active, and we get here, we are in failsafe landing mode, it controls throttle
     if (failsafeIsActive()) {
         // descend at up to 10 times faster when high
-        // default landing time is now 60s; need to get the quad down in this time from reasonable height
-        // need a rapid descent if high to avoid that timeout, and must slow down closer to ground
+        // default landing timeout is now 60s; must to get the quad down within this limit
+        // need a rapid descent when initiated high, and must slow down closer to ground
         // this code doubles descent rate at 20m, to max 10x (10m/s on defaults) at 200m
-        // user should be able to descend within 60s from around 150m high without disarming, on defaults
         // the deceleration may be a bit rocky if it starts very high up
         // constant (set) deceleration target in the last 2m
-        throttleAdjustmentFactor = -(0.9f + constrainf(getAltitudeCm() / 2000.0f, 0.1f, 9.0f));
+        stickFactor = -(0.9f + constrainf(getAltitudeCm() / 2000.0f, 0.1f, 9.0f));
     }
-    altHoldState.targetAltitudeAdjustRate = throttleAdjustmentFactor * altholdConfig()->alt_hold_target_adjust_rate;
+    altHold.targetVelocity = stickFactor * altHold.maxVelocity;
 
-    // if taskRate is 100Hz, default adjustRate of 100 adds/subtracts 1m every second, or 1cm per task run, at full stick position
-    altHoldState.targetAltitudeCm += altHoldState.targetAltitudeAdjustRate  * taskIntervalSeconds;
+    // prevent stick input from moving target altitude too far away from current altitude
+    // otherwise it can be difficult to bring target to the other side of current in reasonable time
```

Comment:
Kind of hard to explain.

Imagine the quad is climbing fast from a pre-existing rapid asscent immediately from takeoff, and the pilot wants to quickly enter altitude hold (or position hold with altitude hold) because they are losing sight of it.  They hit the switch.  The aircraft continues to climb above the initial target altitude due to momentum.  The natural response is to push throttle to zero.  In acro that works, you get less thrust instantly.  But in altitude hold, the PIDs already do this for you; you already have the minimum throttle allowed, so it makes no difference.  What low throttle does achieve is to drive the target point further downwards - at 5m every second the throttle is low.

Finally (can take 3-4 seconds), the quad reaches the apex of it's vertical travel and starts to drop.  Although D slows it somewhat, it keeps falling towards the target, even though it can still be concerningly high.  The pilot may still have throttle low.  By now, so much time has passed (can easily be 6-8 seconds), the target altitude can be close to ground level.  The quad falls as fast as altitude hold will permit towards ground, and only slows down a little just before a hard impact.

This happened to me several times when testing the ability of altitude hold to handle high pre-existing vertical climb rates.

It's why I now also force a zero ascend or descend rate when throttle is zero.  That helps solve this problem.

However, limiting the ability of the pilot to lower the altitude hold setpoint to not a really long way below the current (or initial) altitude, also solves this problem.  It also solves the problem generally, in that for instance if the pilot flies over an updraft, it can be difficult avoiding some ascent, or getting the descent rate you want, and pushing the target altitude lower and lower won't solve the problem.  What will happen once the aircraft clear the updraft, the target altitude will have grown negatively, and now is quite low, so that the aircraft drops rapidly.  And when the stick is moved upwards, the target can only increment at 5m/s.  So effectlively you have little or no control until you get the target back up, and that takes time if you are well below current altitude.

The reverse problem happens in a hard drop, but there the PIDs can raise throttle, and a quadcopter can climb much faster (6-8g's) than it can fall (less than 1g).  So it's not so bad, the throttle usually overshoots and the stop is sufficiently fast that few pilots would raise throttle in addition to what the PIDs in altitude hold would do.

Maybe I can say "bring target altitude close to current altitude in a reasonable time"?